DOC-8057 Added the other pages (#2045)
couchbase/docs-server: modules/n1ql/pages/n1ql-language-reference/objectfun.adoc

> * If [.var]`object` is not an object or NULL, it returns a NULL value object.

Reviewer: Many of these examples will not work with travel-sample out of the box, because they use non-standard data. Please add the following paragraph to the start of this page, after the title but before the first function (OBJECT_ADD).

```asciidoc
If you want to try the examples on this page, you must first run the following query to insert additional data to the `travel-sample` bucket.

WARNING: Please note that the query below will alter the data in your sample buckets.
To restore your sample data, remove and reinstall the `travel-sample` bucket.
Refer to xref:manage:manage-settings/install-sample-buckets.adoc[Sample Buckets] for details.

[source,n1ql]
----
UPDATE `travel-sample`.inventory.route
SET schedule[0] = {"day" : 7, "special_flights": [
                    {"flight": "AI444", "utc": "4:44:44", "codename": "green"},
                    {"flight": "AI333", "utc": "3:33:33"},
                    {"flight": "AI222", "utc": "2:22:22", "codename": "yellow"}
                  ] }
WHERE destinationairport = "CDG" AND sourceairport = "TLV";
----
```

Reviewer: Moved "special flights" update and warning message to top of page

OBJECT_CONCAT(expr1, expr2 ...)
----

.Example
[source,n1ql]
----
SELECT OBJECT_CONCAT({"abc":1},{"def":2},{"fgh":3});

> {

Reviewer: Separate query and results

```suggestion
SELECT OBJECT_CONCAT({"abc":1},{"def":2},{"fgh":3});
----

.Results
[source,json]
----
{
```

Return Value:: An array of values or name-value pairs in N1QL collation order and only the documents for which the parent has a relationship with a child.

*Example 1*: Find one of the flights leaving TLV and arriving at CDG.
[source,n1ql]
----
SELECT OBJECT_INNER_PAIRS(schedule[0])
FROM `travel-sample`.inventory.route
WHERE
    destinationairport = "CDG"
    AND sourceairport = "TLV"
LIMIT 1;
----

Results:
[source,json]
----
[
  {
    "$1": [
      {
        "name": "day",
        "val": 7
      },
      {

>         "name": "special_flights",
>         "val": [
>           {
>             "flight": "AI444",
>             "utc": "4:44:44"
>           },
>           {
>             "flight": "AI333",
>             "utc": "3:33:33"
>           }
>         ]

Reviewer: Update results

```suggestion
        "name": "special_flights",
        "val": [
          {
            "codename": "green",
            "flight": "AI444",
            "utc": "4:44:44"
          },
          {
            "flight": "AI333",
            "utc": "3:33:33"
          },
          {
            "codename": "yellow",
            "flight": "AI222",
            "utc": "2:22:22"
          }
        ]
```

In the example below, one of the elements in the special_flights array does not have a codename and the output of the function contains only two values and does not include the null entry.

.Example
[source,n1ql]
----
SELECT OBJECT_INNER_VALUES(schedule[0].special_flights[*])
FROM `travel-sample`.inventory.route
WHERE destinationairport = "CDG"
    AND sourceairport = "TLV"
LIMIT 1;
----
.Results
[source,json]
----

> [
>   {
>     "$1": [
>       [
>         "AI444",
>         "AI333"
>       ],
>       [
>         "4:44:44",
>         "3:33:33"
>       ]
>     ]
>   }
> ]

Reviewer: Update results

```suggestion
[
  {
    "$1": [
      [
        "green",
        "yellow"
      ],
      [
        "AI444",
        "AI333",
        "AI222"
      ],
      [
        "4:44:44",
        "3:33:33",
        "2:22:22"
      ]
    ]
  }
]
```

In the example below, one of the elements in the special_flights array does not have a codename and the output of the function contains three values, including the null entry.

.Example
[source,n1ql]
----
SELECT OBJECT_PAIRS(schedule[0].special_flights[*])
FROM `travel-sample`.inventory.route
WHERE destinationairport = "CDG"
      AND sourceairport = "TLV"
LIMIT 1;
----
.Results
[source,json]
----

> [
>   {
>     "$1": [
>       {
>         "name": "flight",
>         "val": [
>           "AI444",
>           "AI333"
>         ]
>       },
>       {
>         "name": "utc",
>         "val": [
>           "4:44:44",
>           "3:33:33"
>         ]
>       }
>     ]
>   }
> ]

Reviewer: Update results

```suggestion
[
  {
    "$1": [
      {
        "name": "codename",
        "val": [
          "green",
          null,
          "yellow"
        ]
      },
      {
        "name": "flight",
        "val": [
          "AI444",
          "AI333",
          "AI222"
        ]
      },
      {
        "name": "utc",
        "val": [
          "4:44:44",
          "3:33:33",
          "2:22:22"
        ]
      }
    ]
  }
]
```

*Return Value*:: The JSON object `input_obj` with the new field name.

*Examples*::

> Changing the metadata field name from "name" to be "new_name"

Reviewer: It's not metadata

```suggestion
Changing the field name from "name" to be "new_name"
```

* If the [.var]`attr_key` is not an object, it returns a NULL value.

.Example
[source,n1ql]
----
SELECT OBJECT_REMOVE(schedule[0], "day")
FROM `travel-sample`.inventory.route
LIMIT 1;
----
.Results
[source,json]
----
[
  {
    "$1": {
      "flight": "AF198",
      "utc": "10:13:00"
    }
  }
]
----

> //resulting bad request
> .Example

Reviewer: Seems fine to me!

```suggestion
.Example
```

[source,n1ql]
----

> SELECT OBJECT_REMOVE( {"abc":1,"def":2,"fgh":3},"def");
> {

Reviewer: Separate query and request

```suggestion
SELECT OBJECT_REMOVE( {"abc":1,"def":2,"fgh":3},"def");
----

.Results
[source,json]
----
{
```

*Examples*::
+

> Changing the metadata field name from "name" to be "new_name"

Reviewer: ```suggestion
Changing the `type` field value from "airline" to be "airplane"
```

[source,n1ql]
----

> SELECT object_rename(t, "airline", "airplane")

Reviewer: Wrong function in example!

```suggestion
SELECT object_replace(t, "airline", "airplane")
```

FROM `travel-sample`.inventory.airline AS t
LIMIT 1;
----
.Results
[source,n1ql]
----
[

>   {
>     "$1": {
>       "callsign": "MILE-AIR",
>       "country": "United States",
>       "iata": "Q5",
>       "icao": "MLA",
>       "id": 10,
>       "name": "40-Mile Air",
>       "type": "airline"
>     }
>   }

Reviewer: Update results

```suggestion
  {
    "$1": {
      "callsign": "MILE-AIR",
      "country": "United States",
      "iata": "Q5",
      "icao": "MLA",
      "id": 10,
      "name": "40-Mile Air",
      "type": "airplane"
    }
  }
```

In the example below, one of the elements in the special_flights array does not have a codename and the output of the function contains three values, including the null entry.

.Example
[source,n1ql]
----
SELECT OBJECT_VALUES(schedule[0].special_flights[*])
FROM `travel-sample`.inventory.route
WHERE destinationairport = "CDG"
    AND sourceairport = "TLV"
LIMIT 1;
----
.Results
[source,json]
----

> [
>   {
>     "$1": [
>       [
>         "AI444",
>         "AI333"
>       ],
>       [
>         "4:44:44",
>         "3:33:33"
>       ]
>     ]
>   }
> ]

Reviewer: Update results

```suggestion
[
  {
    "$1": [
      [
        "green",
        null,
        "yellow"
      ],
      [
        "AI444",
        "AI333",
        "AI222"
      ],
      [
        "4:44:44",
        "3:33:33",
        "2:22:22"
      ]
    ]
  }
]
```
